tudatBundle/pagmo2/include/pagmo/algorithms/giACOmo.hpp

```diff
@@ -0,0 +1,960 @@
+/// Extended ACO
+/**
+ * \image html ACO.jpg "The ACO flowchart" width=3cm [TO BE ADDED]
+
+ * ACO is inspired by the natural mechanism with which real ant colonies forage food.
+ * This algorithm has shown promising results in many trajectory optimization problems.
+ * The first appearance of the algorithm happened in Dr. Marco Dorigo's thesis, in 1992.
+ * ACO generates future generations of ants by using the a multi-kernel gaussian distribution
+ * based on three parameters (i.e., pheromone values) which are computed depending on the quality
+ * of each previous solution. The solutions are ranked through an oracle penalty method.
+ *
+ *
+ * The version implemented in pagmo can be applied to box-bounded multiple-objective optimization.
+ *
+ * See:  M. Schlueter, et al. (2009). Extended ant colony optimization for non-convex
+ * mixed integer non-linear programming. Computers & Operations Research.
+ */
+class gi_aco_mo
+{
+public:
+    /// Single entry of the log (gen, fevals, ideal_point)
+    typedef std::tuple<unsigned, unsigned long long, vector_double> log_line_type;
+    /// The log
+    typedef std::vector<log_line_type> log_type;
+
+    /// Constructors
+    /**
+    * Constructs the ACO user defined algorithm for single-objective optimization.
+    *
+    * @param[in] gen Generations: number of generations to evolve.
+    * @param[in] acc Accuracy parameter: for inequality and equality constraints .
+    * @param[in] fstop Objective stopping criterion: when the objective value reaches this value, the algorithm is stopped [for multi-objective, this applies to the first obj. only].
+    * @param[in] impstop Improvement stopping criterion: if a positive integer is assigned here, the algorithm will count the runs without improvements, if this number will exceed IMPSTOP value, the algorithm will be stopped.
+    * @param[in] evalstop Evaluation stopping criterion: same as previous one, but with function evaluations
+    * @param[in] focus Focus parameter: this parameter makes the search for the optimum greedier and more focused on local improvements (the higher the greedier). If the value is very high, the search is more focused around the current best solutions
+    * @param[in] ker Kernel: number of solutions stored in the solution archive
+    * @param[in] oracle Oracle parameter: this is the oracle parameter used in the penalty method
+    * @param seed seed used by the internal random number generator (default is random)
+    * @throws std::invalid_argument if \p acc is not \f$ \in [0,1[\f$, \p fstop is not positive, \p impstop is not a
+    * positive integer, \p evalstop is not a positive integer, \p focus is not \f$ \in [0,1[\f$, \p ants is not a positive integer,
+    * \p ker is not a positive integer, \p oracle is not positive, \p paretomax is not a positive integer, \p epsilon is not \f$ \in [0,1[\f$
+    */
+    gi_aco_mo(unsigned gen = 1u, double acc = 0.95, double  fstop = 1, unsigned impstop = 1, unsigned evalstop = 1,
```

Comment:
What is the rationale for having two different constructors for SO vs MO optimisation? And what happens if a MO ACO is used to evolve a SO problem?

The logic for differentiating between SO and MO should go in the ``evolve()`` method, if possible.

Reply:
In the SO case I do not need the PARETOMAX and EPSILON parameters, so I thought about doing two different constructors. Are there disadvantages?
I still cannot exhaustively answer the second question since I have not defined the MO yet, but I guess that I would throw if the user inserts as many inputs as requested by the MO constructor but it inputs only one objective function.

About this I also had a question: in the NSGA-II there is the check that prob.get_nf() < 2u, and if this is verified, the algorithm throws. However, get_nf() would return also the number of ec and nec. Shouldn't it be: (prob.get_nobj() < 2u ) ?

By the way, I was thinking to implement a similar check in my case also (I have an if statement that divides the SO from the MO inside the evolve, so this should be easy to do).

Reply:
I would leave a single constructor with all the possible parameters, and document that the PARETOMAX and EPSILON are relevant only for MO optimisation (although I would not error out if the user specifies them for a SO optimisation as well). Perhaps you can put some default values for them, so the user who wants only SO optimisation does not have to specify them?

Note that in pagmo we also normally require algorithms to be default-constructible, so you will need to have a constructor without parameters as well.

The problem with the current approach is that you have 2 different places where the difference between MO and SO matters (the constructor and the evolve), so you have 4 possible states you need to deal with in the algorithm.

Reply:
Regarding the NSGA2 question, it seems to me like the code is checking before that there are no constraintsm, so it gets to the get_nf() < 2u only in absence of constraints (which means get_nf() == get_nobj()). @darioizzo ?

Reply:
Indeed, did we not agree to only have a single objective version to start and then deal with the extensions?

Reply:
Indeed, did we not agree to only have a single objective version to start and then deal with the extensions?

Reply:
I think this would be the preferred plan as it would make the whole project and its code supervision much more easy. 

Also it would be easier to tackle the issues one at a time.

Reply:
Okay, thanks for the advice. I will switch to one single constructor and use only the two if statements in the evolve for distinguishing MO from SO

Reply:
@darioizzo 
Do you mean that I should do two different .hpp files for SO and MO, or is it okay to keep the if statements?

Reply:
I mean that the first task for you is to develop a single objective algorithm. Do not add any code that relates to a multi-objective version. We will deal with that once the single objective work has finished.

Reply:
Yes, sure. 

I think that the main part of the SO ACO is completed, I now have to compile it to correct possible compile errors.

Reply:
That is good news, after the c++ code is in good shape, the second step is to have C++ tests of the code. Then we "expose" it to python, then we see how it behaves on test roblems (rosenbrock, etc...)

Reply:
The C++ code compiles, shall I now handle the log, or shall I immediately start with the tests?